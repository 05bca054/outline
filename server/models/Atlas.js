import {
  DataTypes,
  sequelize,
} from '../sequelize';
import _isEqual from 'lodash/isEqual';
import Document from './Document';

const allowedAtlasTypes = [['atlas', 'journal']];

const Atlas = sequelize.define('atlas', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  name: DataTypes.STRING,
  description: DataTypes.STRING,
  type: { type: DataTypes.STRING, validate: { isIn: allowedAtlasTypes }},

  /* type: atlas */
  navigationTree: DataTypes.JSONB,
}, {
  tableName: 'atlases',
  hooks: {
  //   beforeValidate: (doc) => {
  //     doc.urlId = randomstring.generate(15);
  //   },
  //   beforeCreate: (doc) => {
  //     doc.html = convertToMarkdown(doc.text);
  //     doc.preview = truncateMarkdown(doc.text, 160);
  //   },
  //   beforeUpdate: (doc) => {
  //     doc.html = convertToMarkdown(doc.text);
  //     doc.preview = truncateMarkdown(doc.text, 160);
  //   },
  },
  instanceMethods: {
    async getStructure() {
      if (this.navigationTree) {
        return this.navigationTree;
      }

      const getNodeForDocument = async (document) => {
        const children = await Document.findAll({ where: {
          parentDocumentId: document.id,
          atlasId: this.id,
        }});

        let childNodes = []
        await Promise.all(children.map(async (child) => {
          childNodes.push(await getNodeForDocument(child));
        }));

        return {
          title: document.title,
          id: document.id,
          url: document.getUrl(),
          children: childNodes,
        };
      }

      const rootDocument = await Document.findOne({
        where: {
          parentDocumentId: null,
          atlasId: this.id,
        }
      });

      if (rootDocument) {
        return await getNodeForDocument(rootDocument);
      } else {
        return; // TODO should create a root doc
      }
    },
    async updateNavigationTree(tree = this.navigationTree) {
      let nodeIds = [];
      nodeIds.push(tree.id);

      const rootDocument = await Document.findOne({
        where: {
          id: tree.id,
          atlasId: this.id,
        },
      });
      if (!rootDocument) throw new Error;

      let newTree = {
        id: tree.id,
        title: rootDocument.title,
        url: rootDocument.getUrl(),
        children: [],
      };

      const getIdsForChildren = async (children) => {
        const childNodes = [];
        for (const child of children) {
          const childDocument = await Document.findOne({
            where: {
              id: child.id,
              atlasId: this.id,
            },
          });
          if (childDocument) {
            childNodes.push({
              id: childDocument.id,
              title: childDocument.title,
              url: childDocument.getUrl(),
              children: await getIdsForChildren(child.children),
            })
            nodeIds.push(child.id);
          }
        }
        return childNodes;
      };
      newTree.children = await getIdsForChildren(tree.children);

      const documents = await Document.findAll({
        attributes: ['id'],
        where: {
          atlasId: this.id,
        }
      });
      const documentIds = documents.map(doc => doc.id);

      if (!_isEqual(nodeIds.sort(), documentIds.sort())) {
        throw new Error('Invalid navigation tree');
      }

      this.navigationTree = newTree;
      await this.save();

      return newTree;
    },
    async addNodeToNavigationTree(document) {
      const newNode = {
        id: document.id,
        title: document.title,
        url: document.getUrl(),
        children: [],
      }

      const insertNode = (node) => {
        if (document.parentDocumentId === node.id) {
          node.children.push(newNode);
        } else {
          node.children = node.children.map(childNode => {
            return insertNode(childNode);
          })
        }

        return node;
      };

      this.navigationTree = insertNode(this.navigationTree);
    },
    async deleteDocument(document) {
      const deleteNodeAndDocument = async (node, documentId, shouldDelete = false) => {
        if (document.id === node.id) {
          shouldDelete = true;
        }
        const newChildren = [];
        node.children.map(async childNode => {
          const child = await deleteNodeAndDocument(childNode, documentId, shouldDelete);
          if (child) {
            newChildren.push(child);
          }
        });
        node.children = newChildren;

        if (shouldDelete) {
          const document = await Document.findById(node.id);
          await document.destroy();
        }

        return shouldDelete ? null : node;
      };

      this.navigationTree = await deleteNodeAndDocument(this.navigationTree, document.id);
    }
  }
});

Atlas.hasMany(Document, { as: 'documents', foreignKey: 'atlasId' });

export default Atlas;
