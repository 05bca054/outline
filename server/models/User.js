import crypto from 'crypto';
import {
  DataTypes,
  sequelize,
  encryptedFields
} from '../sequelize';
import Team from './Team';

import JWT from 'jsonwebtoken';

const User = sequelize.define('user', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  email: DataTypes.STRING,
  username: DataTypes.STRING,
  name: DataTypes.STRING,
  isAdmin: DataTypes.BOOLEAN,
  slackAccessToken: encryptedFields.vault('slackAccessToken'),
  slackId: { type: DataTypes.STRING, unique: true },
  slackData: DataTypes.JSONB,
  jwtSecret: encryptedFields.vault('jwtSecret'),
}, {
  instanceMethods: {
    getJwtToken() {
      return JWT.sign({ id: this.id }, this.jwtSecret);
    },
  },
  indexes: [
    {
      unique: true,
      fields: ['email']
    },
  ],
});

const setRandomJwtSecret = (model) => {
  model.jwtSecret = crypto.randomBytes(64).toString('hex');
};

User.beforeCreate(setRandomJwtSecret);
User.belongsTo(Team);

sequelize.sync();

export default User;
