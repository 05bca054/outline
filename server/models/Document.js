import slug from 'slug';
import randomstring from 'randomstring';
import {
  DataTypes,
  sequelize,
} from '../sequelize';
import {
  convertToMarkdown,
  truncateMarkdown,
} from '../utils/markdown';
import Atlas from './Atlas';
import Team from './Team';
import User from './User';

slug.defaults.mode ='rfc3986';

const generateSlug = (title, urlId) => {
  const slugifiedTitle = slug(title);
  return `${slugifiedTitle}-${urlId}`;
};

const Document = sequelize.define('document', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  urlId: { type: DataTypes.STRING, primaryKey: true },
  private: { type: DataTypes.BOOLEAN, defaultValue: true },
  title: DataTypes.STRING,
  text: DataTypes.TEXT,
  html: DataTypes.TEXT,
  preview: DataTypes.TEXT,
}, {
  hooks: {
    beforeValidate: (doc) => {
      doc.urlId = randomstring.generate(15);
    },
    beforeCreate: (doc) => {
      doc.html = convertToMarkdown(doc.text);
      doc.preview = truncateMarkdown(doc.text, 160);
    },
    beforeUpdate: (doc) => {
      doc.html = convertToMarkdown(doc.text);
      doc.preview = truncateMarkdown(doc.text, 160);
    },
  },
  instanceMethods: {
    buildUrl() {
      const slugifiedTitle = slug(this.title);
      return `${slugifiedTitle}-${this.urlId}`;
    }
  }
});

Document.belongsTo(Atlas, { as: 'atlas' });
Document.belongsTo(Team);
Document.belongsTo(User);

export default Document;
