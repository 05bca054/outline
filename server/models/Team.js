import {
  DataTypes,
  sequelize,
} from '../sequelize';
import Atlas from './Atlas';
import Document from './Document';
import User from './User';

const Team = sequelize.define('team', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  name: DataTypes.STRING,
  slackId: { type: DataTypes.STRING, unique: true },
  slackData: DataTypes.JSONB,
}, {
  instanceMethods: {
    async createFirstAtlas() {
      const atlas = await Atlas.create({
        name: this.name,
        description: 'Your first Atlas',
        type: 'journal',
        teamId: this.id,
      });
      return atlas;
    }
  },
  indexes: [
    {
      unique: true,
      fields: ['slackId']
    },
  ],
});

Team.hasMany(Atlas, { as: 'atlases' });
Team.hasMany(Document, { as: 'documents' });
Team.hasMany(User, { as: 'users' });

export default Team;
