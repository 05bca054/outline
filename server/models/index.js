import User from './User';
import Team from './Team';
import Atlas from './Atlas';
import Document from './Document';
import Revision from './Revision';

export {
  User,
  Team,
  Atlas,
  Document,
  Revision,
};
