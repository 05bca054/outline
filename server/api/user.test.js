import TestServer from 'fetch-test-server';

import app from '..';
import { User } from '../models';

import { flushdb, seed, sequelize } from '../test/support';

const server = new TestServer(app.callback());

beforeEach(flushdb);
beforeEach(seed);
afterAll(() => server.close());
afterAll(() => sequelize.close());

it('should return known user', async () => {
  const user = await User.findOne({
    where: {
      email: 'user1@example.com',
    },
  });

  const res = await server.post('/api/user.info', {
    body: { token: user.getJwtToken() },
  });
  const body = await res.json();

  expect(res.status).toEqual(200);
  expect(body).toMatchSnapshot();
});

it('should require authentication', async () => {
  const res = await server.post('/api/user.info');
  const body = await res.json();

  expect(res.status).toEqual(401);
  expect(body).toMatchSnapshot();
});
