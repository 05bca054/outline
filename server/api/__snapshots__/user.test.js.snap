exports[`test should require authentication 1`] = `
Object {
  "message": "Authentication required"
}
`;

exports[`test should return known user 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "id": "86fde1d4-0050-428f-9f0b-0bf77f8bdf61",
    "name": "User 1",
    "username": "user1"
  }
}
`;
